src/main/java/ratpack/zipkin/internal/ZipkinHttpClientImpl.java

```diff
@@ -111,65 +113,165 @@ public void close() {
+        handler.handleReceive(statusCode, response.getThrowable(), span);
     }
 
-    void resultWithSpan(Result<?> result, StatusCode statusCode, Span span) {
-        if (result.isError()) {
-            String message = result.getThrowable().getMessage();
-            if (message != null) {
-                span.tag(Constants.ERROR, result.getThrowable().getClass().getSimpleName());
-            }
-        } else {
-            int status = statusCode.getStatusCode();
-            if (status < 200 || status > 299) {
-                span.tag(TraceKeys.HTTP_STATUS_CODE, String.valueOf(status));
-            }
-            if (status > 399) {
-                span.tag(Constants.ERROR, "server error " + status);
-            }
-        }
-        span.finish();
+    private void responseWithSpan(Result<ReceivedResponse> response, AtomicReference<Span> ref) {
+        Span span = ref.get();
+        if (span == null) return;
+
+        Integer statusCode = (response.isError() || response.getValue() == null)
+            ? null : response.getValue().getStatusCode();
+
+        handler.handleReceive(statusCode, response.getThrowable(), span);
     }
 
-    private interface StatusCode {
-        int getStatusCode();
+    static final class HttpAdapter
+        extends brave.http.HttpClientAdapter<WrappedRequestSpec, Integer> {
+
+        @Override public String method(WrappedRequestSpec request) {
+            HttpMethod method = Optional.ofNullable(request.getCapturedMethod()).orElse(HttpMethod.GET);
+            return method.getName();
+        }
+
+        @Override public String path(WrappedRequestSpec request) {
+            return request.getUri().getPath();
+        }
+
+        @Override public String url(WrappedRequestSpec request) {
+            return request.getUri().toString();
+        }
+
+        @Override public String requestHeader(WrappedRequestSpec request, String name) {
+            return request.getHeaders().get(name);
+        }
+
+        @Override public Integer statusCode(Integer response) {
+            return response;
+        }
     }
 
-    private static final class DefaultRequestSpecNameAdapter implements
-        SpanNameProvider.SpanNameProviderAdapter {
-        private final MethodCapturingRequestSpec request;
+    /**
+     * RequestSpec wrapper that captures the method type, sets up redirect handling
+     * and starts new spans when a method type is set.
+     */
+    static final class WrappedRequestSpec implements RequestSpec {
 
-        DefaultRequestSpecNameAdapter(MethodCapturingRequestSpec request) {
-            this.request = request;
+        private final RequestSpec delegate;
+        private final ZipkinHttpClientImpl client;
+        private final AtomicReference<Span> span;
+        private HttpMethod capturedMethod;
+
+        WrappedRequestSpec(ZipkinHttpClientImpl client, RequestSpec spec,  AtomicReference<Span> span) {
```

Comment:
@llinder what do you think about just passing just what this class needs (i.e. handler and injector) instead of the whole client?